// npm-upgrade/lib/packageUtils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadPackageJson = loadPackageJson;
exports.findModuleDepsGroup = findModuleDepsGroup;
exports.getModuleVersion = getModuleVersion;
exports.setModuleVersion = setModuleVersion;
exports.getModuleHomepage = getModuleHomepage;
exports.getModuleInfo = exports.DEPS_GROUPS = void 0;

var _memoize2 = _interopRequireDefault(require("lodash/memoize"));

var _path = require("path");

var _fs = require("fs");

var _libnpmconfig = _interopRequireDefault(require("libnpmconfig"));

var _pacote = _interopRequireDefault(require("pacote"));

const DEPS_GROUPS = [{
  name: 'production',
  field: 'dependencies',
  flag: 'p',
  ncuValue: 'prod'
}, {
  name: 'optional',
  field: 'optionalDependencies',
  flag: 'o',
  ncuValue: 'optional'
}, {
  name: 'development',
  field: 'devDependencies',
  flag: 'd',
  ncuValue: 'dev'
}, {
  name: 'peer',
  field: 'peerDependencies',
  flag: 'r',
  ncuValue: 'peer'
}, {
  name: 'bundled',
  field: 'bundledDependencies',
  altField: 'bundleDependencies',
  flag: 'b',
  ncuValue: 'bundle'
}];
exports.DEPS_GROUPS = DEPS_GROUPS;
const getNpmConfig = (0, _memoize2.default)(() => {
  const config = {};

  _libnpmconfig.default.read().forEach((value, key) => {
    if (typeof value === 'string') {
      // Replacing env ${VARS} in strings with the `process.env` values
      config[key] = value.replace(/\$\{([\0-\t\x0B\f\x0E-\u2027\u202A-\u{10FFFF}]+?)\}/gu, (_, envVar) => process.env[envVar]);
    } else {
      config[key] = value;
    }
  });

  return config;
});

function loadPackageJson() {
  const packageFile = (0, _path.resolve)('./package.json');
  let packageJson;
  let packageSource;

  try {
    packageSource = (0, _fs.readFileSync)(packageFile, 'utf-8');
    packageJson = JSON.parse(packageSource);
  } catch (err) {
    console.error(`Error loading package.json: ${err.message}`);
    process.exit(1);
  }

  return {
    path: packageFile,
    content: packageJson,
    source: packageSource
  };
}

function findModuleDepsGroup(moduleName, packageJson) {
  for (const {
    field,
    altField
  } of DEPS_GROUPS) {
    let modules = packageJson[field];

    if (!modules && altField) {
      modules = packageJson[altField];
    }

    if (modules && modules[moduleName]) {
      return modules;
    }
  }

  return null;
}

function getModuleVersion(moduleName, packageJson) {
  const depsGroup = findModuleDepsGroup(moduleName, packageJson);
  return depsGroup ? depsGroup[moduleName] : null;
}

function setModuleVersion(moduleName, newVersion, packageJson) {
  const depsGroup = findModuleDepsGroup(moduleName, packageJson);

  if (depsGroup) {
    depsGroup[moduleName] = newVersion;
    return true;
  } else {
    return false;
  }
}

function getModuleHomepage(packageJson) {
  return packageJson.homepage || packageJson.url || null;
}

const getModuleInfo = (0, _memoize2.default)(async moduleName => _pacote.default.manifest(moduleName, { ...getNpmConfig(),
  'full-metadata': true
}));
exports.getModuleInfo = getModuleInfo;
